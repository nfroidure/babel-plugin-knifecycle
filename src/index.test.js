// src/__tests__/index-test.js
import { transform } from '@babel/core';
import plugin from '.';

describe('babel-plugin-knifecycle', () => {
  describe('autoInject', () => {
    it('should work with es6 modules imports', () => {
      var example = `
        import noop from 'noop';
        import { autoInject } from 'knifecycle';

        export default autoInject(getUser);

        async function getUser({ mysql: db, log = noop }, { userId }) {
          return {};
        }
        `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should work with no injection', () => {
      var example = `
        import noop from 'noop';
        import { autoInject } from 'knifecycle';

        export default autoInject(getUser);

        async function getUser() {
          return {};
        }
        `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should work with several auto functions', () => {
      var example = `
        import noop from 'noop';
        import { autoInject, autoName } from 'knifecycle';

        export default autoInject(autoName(getUser));

        async function getUser({ mysql: db, log = noop }, { userId }) {
          return {};
        }
        `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });
  });

  describe('autoName', () => {
    it('should work with es6 modules imports', () => {
      var example = `
        import noop from 'noop';
        import { autoName } from 'knifecycle';

        export default autoName(getUser);

        async function getUser({ mysql: db, log = noop }, { userId }) {
          return {};
        }
        `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should work with init prefix', () => {
      var example = `
        import noop from 'noop';
        import { autoName } from 'knifecycle';

        export default autoName(initGetUser);

        async function initGetUser({ mysql: db, log = noop }, { userId }) {
          return {};
        }
        `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });
  });

  describe('autoService', () => {
    it('should work with es6 modules imports', () => {
      var example = `
          import noop from 'noop';
          import { autoService } from 'knifecycle';

          export default autoService(getUser);

          async function getUser({ mysql: db, log = noop }) {
            return async ({ userId }) => {
              return {};
            }
          }
          `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should work with no dependencies', () => {
      var example = `
          import noop from 'noop';
          import { autoService } from 'knifecycle';

          export default autoService(getUser);

          async function getUser() {
            return async ({ userId }) => {
              return {};
            }
          }
          `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });
  });

  describe('autoProvider', () => {
    it('should work with es6 modules imports', () => {
      var example = `
          import noop from 'noop';
          import { autoProvider } from 'knifecycle';

          export default autoProvider(getUser);

          async function getUser({ mysql: db, log = noop }) {
            return async ({ userId }) => {
              return {};
            }
          }
          `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should work with options', () => {
      var example = `
          import noop from 'noop';
          import { autoProvider } from 'knifecycle';

          export default autoProvider(getUser, { singleton: true });

          async function getUser({ mysql: db, log = noop }) {
            return async ({ userId }) => {
              return {};
            }
          }
          `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });
  });

  describe('autoHandler', () => {
    it('should work with es6 modules imports', () => {
      var example = `
          import noop from 'noop';
          import { autoHandler } from 'knifecycle';

          export default autoHandler(getUser);

          async function getUser({ mysql: db, log = noop }, { userId }) {
            return {};
          }
          `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should work with inline function expressions', () => {
      var example = `
      import noop from 'noop';
      import { autoHandler } from 'knifecycle';

      export default autoHandler(async function getUser({ mysql: db, log = noop }, { userId }) {
        return {};
      });
      `;

      const { code } = transform(example, { plugins: [plugin] });
      expect(code).toMatchSnapshot();
    });

    it('should fail with anonymous function expressions', () => {
      var example = `
      import noop from 'noop';
      import { autoHandler } from 'knifecycle';

      export default autoHandler(async ({ mysql: db, log = noop }, { userId }) => {
        return {};
      });
      `;

      expect(() => {
        transform(example, { plugins: [plugin] });
      }).toThrow('Expect autoHandler to take an async function in argument.');
    });

    it('should fail with something defined elsewhere', () => {
      var example = `
      import noop from 'noop';
      import { autoHandler } from 'knifecycle';
      import getUser from './getUser';

      export default autoHandler(getUser);
      `;

      expect(() => {
        transform(example, { plugins: [plugin] });
      }).toThrow(
        'Expect the function passed in autoHandler to be defined in the local file.',
      );
    });
  });
});
