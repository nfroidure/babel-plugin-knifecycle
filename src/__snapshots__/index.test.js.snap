// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-knifecycle autoHandler should work with es6 modules imports 1`] = `
"import noop from 'noop';
import { handler as _handler } from 'knifecycle';
export default _handler(getUser, \\"getUser\\", [\\"mysql\\", \\"?log\\"]);

async function getUser({
  mysql: db,
  log = noop
}, {
  userId
}) {
  return {};
}"
`;

exports[`babel-plugin-knifecycle autoHandler should work with inline function expressions 1`] = `
"import noop from 'noop';
import { handler as _handler } from 'knifecycle';
export default _handler(async function getUser({
  mysql: db,
  log = noop
}, {
  userId
}) {
  return {};
}, \\"getUser\\", [\\"mysql\\", \\"?log\\"]);"
`;

exports[`babel-plugin-knifecycle autoInject should work with es6 modules imports 1`] = `
"import noop from 'noop';
import { inject as _inject } from 'knifecycle';
export default _inject([\\"mysql\\", \\"?log\\"], getUser);

async function getUser({
  mysql: db,
  log = noop
}, {
  userId
}) {
  return {};
}"
`;

exports[`babel-plugin-knifecycle autoInject should work with several auto functions 1`] = `
"import noop from 'noop';
import { inject as _inject, name as _name } from 'knifecycle';
export default _inject([\\"mysql\\", \\"?log\\"], _name(\\"getUser\\", getUser));

async function getUser({
  mysql: db,
  log = noop
}, {
  userId
}) {
  return {};
}"
`;

exports[`babel-plugin-knifecycle autoName should work with es6 modules imports 1`] = `
"import noop from 'noop';
import { name as _name } from 'knifecycle';
export default _name(\\"getUser\\", getUser);

async function getUser({
  mysql: db,
  log = noop
}, {
  userId
}) {
  return {};
}"
`;
